File: frontend/src/lib/cost.ts
import { useQuery } from "@tanstack/react-query";
import { apiFetch } from "./api";

export interface CostByProfile {
  profile: string;
  cost_usd: number;
  turns: number;
  runs: number;
}

export interface CostByState {
  state: string;
  cost_usd: number;
  runs: number;
}

export interface CostByDay {
  date: string;
  cost_usd: number;
}

export interface CostByTicket {
  ticket_key: string;
  cost_usd: number;
  turns: number;
  runs: number;
  state: string;
  updated_at: string;
}

export interface TurnsBucket {
  bucket: string;
  count: number;
}

export interface BudgetStatus {
  monthly_budget_usd: number | null;
  month_spent_usd: number;
  remaining_usd: number | null;
  pct_used: number | null;
}

export interface CostSummary {
  total_cost_usd: number;
  ticket_count: number;
  run_count: number;
  avg_per_ticket_usd: number;
  budget: BudgetStatus;
  by_profile: CostByProfile[];
  by_state: CostByState[];
  by_day: CostByDay[];
  by_ticket: CostByTicket[];
  turns_histogram: TurnsBucket[];
}

export function useCostSummary() {
  return useQuery({
    queryKey: ["cost", "summary"] as const,
    queryFn: () => apiFetch<CostSummary>("/api/cost/summary"),
    refetchInterval: 10_000,
  });
}

File: backend/src/task_summoner/api/schemas/cost.py
"""Cost & usage response schemas."""

from __future__ import annotations

from pydantic import BaseModel


class CostByProfile(BaseModel):
    profile: str
    cost_usd: float
    turns: int
    runs: int


class CostByState(BaseModel):
    state: str
    cost_usd: float
    runs: int


class CostByDay(BaseModel):
    date: str
    cost_usd: float


class CostByTicket(BaseModel):
    ticket_key: str
    cost_usd: float
    turns: int
    runs: int
    state: str
    updated_at: str


class TurnsBucket(BaseModel):
    bucket: str
    count: int


class BudgetStatus(BaseModel):
    monthly_budget_usd: float | None
    month_spent_usd: float
    remaining_usd: float | None
    pct_used: float | None


class CostSummaryResponse(BaseModel):
    total_cost_usd: float
    ticket_count: int
    run_count: int
    avg_per_ticket_usd: float
    budget: BudgetStatus
    by_profile: list[CostByProfile]
    by_state: list[CostByState]
    by_day: list[CostByDay]
    by_ticket: list[CostByTicket]
    turns_histogram: list[TurnsBucket]


__all__ = [
    "BudgetStatus",
    "CostByDay",
    "CostByProfile",
    "CostByState",
    "CostByTicket",
    "CostSummaryResponse",
    "TurnsBucket",
]
